src/editors/documentController.ts

```diff
@@ -0,0 +1,228 @@
+        `no longer connected to ${connectionName}`
+      );
+    }
+
+    const find = util.promisify(dataservice.find.bind(dataservice));
+
+    try {
+      const documents = await find(
+        namespace,
+        {
+          _id: documentId
+        },
+        {
+          limit: 1
+        }
+      );
+
+      if (!documents || documents.length === 0) {
+        return null;
+      }
+
+      return JSON.parse(EJSON.stringify(documents[0]));
+    } catch (error) {
+      return this._fetchDocumentFailed(error.message);
+    }
+  }
+
+  async _replaceDocument(data: {
+    documentId: EJSON.SerializableTypes;
+    namespace: string;
+    connectionName: string;
+    newDocument: EJSON.SerializableTypes;
+  }): Promise<void> {
+    log.info('replace document in MongoDB', data);
+
+    const { documentId, namespace, connectionName, newDocument } = data;
+    const dataservice = this._connectionController.getActiveDataService();
+
+    if (dataservice === null) {
+      return this._saveDocumentFailed(
+        `no longer connected to '${connectionName}'`
+      );
+    }
+
+    const findOneAndReplace = util.promisify(
+      dataservice.findOneAndReplace.bind(dataservice)
+    );
+
+    try {
+      await findOneAndReplace(
+        namespace,
+        {
+          _id: documentId
+        },
+        newDocument,
+        {
+          returnOriginal: false
+        }
+      );
+
+      this._telemetryController.trackDocumentUpdated('treeview', true);
+    } catch (error) {
+      return this._saveDocumentFailed(error.message);
+    }
+  }
+
+  async saveMongoDBDocument(data: {
+    documentId: EJSON.SerializableTypes;
+    namespace: string;
+    connectionId: string;
+    newDocument: EJSON.SerializableTypes;
+  }): Promise<void> {
+    const { documentId, namespace, connectionId, newDocument } = data;
+    const activeConnectionId = this._connectionController.getActiveConnectionId();
+    const connectionName = this._connectionController.getSavedConnectionName(
+      connectionId
+    );
+
+    if (activeConnectionId !== connectionId) {
+      return this._saveDocumentFailed(
+        `no longer connected to '${connectionName}'`
+      );
+    }
+
+    this._statusView.showMessage('Saving document...');
+    await this._replaceDocument({
+      documentId,
+      namespace,
+      connectionName,
+      newDocument
+    });
+    this._statusView.hideMessage();
+  }
+
+  async openMongoDBDocument(data: {
+    documentId: EJSON.SerializableTypes;
+    namespace: string;
+  }): Promise<vscode.Uri> {
+    const { documentId, namespace } = data;
+    const activeConnectionId = this._connectionController.getActiveConnectionId();
+    let connectionName = '';
+
+    if (activeConnectionId) {
+      connectionName = this._connectionController.getSavedConnectionName(
+        activeConnectionId
+      );
+    }
+
+    this._statusView.showMessage('Fetching document...');
+    const document = await this._fetchDocument({
```

Comment:
I think we might want to check to make sure this document still exists here and show an error if not. For example when a user disconnects from their connection and then clicks the `Edit Document` code lens it currently shows a bit of a cryptic error which we could handle here I think.

Reply:
Looks like `_fetchDocument` throws an error when the document doesn't exist, so we can catch that here?

Reply:
* chatted - we can move that catching logic into the editors controller.

Reply:
It is already in editorsController, but let's move as we discussed the uri building to editorsController and rename some methods in documentController to resolve this confusion.